sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -684,6 +685,24 @@ object TypeCoercion {
     }
   }
 
+  /**
+   * Since SPARK-16323 `IntegralDivide` always returns long-type value.
+   * This rule cast the integral inputs to long type, to avoid overflow during calculation.
+   */
+  object IntegralDivision extends TypeCoercionRule {
+    override protected def coerceTypes(
+        plan: LogicalPlan): LogicalPlan = plan resolveExpressions {
+      case e if !e.childrenResolved => e
+      case d @ IntegralDivide(left, right) =>
```

Comment:
Let's update `IntegralDivide.inputType`, to only accpet `LongType` and `DecimalType`